### PR TITLE
Report the correct print location in the console

### PR DESCRIPTION
**Pull request by davep**
~~Fixes https://github.com/Textualize/textual/issues/3237~~

~~I'm undecided about this. It doesn't feel quite right to special-case this, but on the other hand `print` is a form of logging that seems to work differently in Textual vs the normal logging facility (which seems to do its own dance to handle the call location, before it makes it to devtools).~~

~~This does correct the issue, this does ensure that `print` reports the location the same as any `self.log` call. So this is a commit made to solicit more detail on the design decisions in the core of Textual.~~

---

Sets us up to fix https://github.com/Textualize/textual/issues/3237

This PR allows the caller to pass in a frame that should be considered to be the current frame, for the purposes of working out the location of the original log/print. This change alone won't fix https://github.com/Textualize/textual/issues/3237 but provides what's needed for a change to Textual that will allow fixing it.

**IMPORTANT:** If this is an acceptable approach, this code should be merged into `textual-dev` and published first. There is [a corresponding PR for Textual itself](https://github.com/Textualize/textual/pull/3294) that should only be merged and used *after* this version of `textual-dev` is available and able to be used. This PR is backward-compatible so should not cause problems for any existing developer-installations of Textual.

### CODE REVIEW

**Comment by davep** [comment]
Throwing out a review request rather than just charging ahead with merging as this is a slightly hacky fix.

**Comment by davep** [comment]
> I think the way I've dealt with similar issues in the past, is to add a parameter for the offset in to the callstack.

That was going to be my initial approach too, but that meant changing the signature of `StdoutRediector.write` (albeit in a backward-compatible way), which felt a little heavy-handed for this one particular case. If we're happy with that I'll look to taking that approach instead; it should end up being easier to maintain.

**Comment by davep** [comment]
Right, done. Note the change to the description of this PR. Also note that this does *not* fix https://github.com/Textualize/textual/issues/3237 but it does provide what's needed to fix it within Textual. Once this is live the fix will be as simple as:

```diff
diff --git a/src/textual/app.py b/src/textual/app.py
index 2991f51e..5634d8f6 100644
--- a/src/textual/app.py
+++ b/src/textual/app.py
@@ -1161,7 +1161,8 @@ class App(Generic[ReturnType], DOMNode):
             stderr: True if the print was to stderr, or False for stdout.
         """
         if self._devtools_redirector is not None:
-            self._devtools_redirector.write(text)
+            current_frame = inspect.currentframe()
+            self._devtools_redirector.write(text, current_frame.f_back if current_frame is not None else None)
         for target, (_stdout, _stderr) in self._capture_print.items():
             if (_stderr and stderr) or (_stdout and not stderr):
                 target.post_message(events.Print(text, stderr=stderr))
```

**Comment by davep** [comment]
Tagging @darrenburns as just suggested by @willmcgugan

**Review by darrenburns — APPROVED** [review]

File: src/textual_dev/tools/diagnose.py [hunk 1]
```diff
@@ -158,7 +158,9 @@ def _console() -> None:
 
 def diagnose() -> None:
     """Print information about Textual and its environment to help diagnose problems."""
-    print("<!-- This is valid Markdown, please paste the following directly in to a GitHub issue -->")
+    print(
+        "<!-- This is valid Markdown, please paste the following directly in to a GitHub issue -->"
```
Comment (rodrigogiraoserrao) [inline src/textual_dev/tools/diagnose.py:162]: Given that this was edited, maybe we can also fix this typo? :P

```suggestion
        "<!-- This is valid Markdown, please paste the following directly into a GitHub issue -->"
```

Comment (davep) [reply src/textual_dev/tools/diagnose.py:162]: It wasn't really edited, just black doing its thing I believe. 

Perhaps PR the change given this PR is already merged and released?

**Pull request merged**